File: soungmunkim/C++/String/132_PalindromePartitioning2.cpp
/*
[palindrome partitions 중에 가장 적은 cut 수 구하기]

Given a string s, 
partition s such that every substring of the partition is a palindrome.

Return the minimum cuts needed for a palindrome partitioning of s.

(base) soungmunkim@gimseongmun-ui-MacBookPro String % ./test                                                
Please enter a string: 
"aab"
["a", "a", "b"], ["aa", "b"]
min cut: 1

(base) soungmunkim@gimseongmun-ui-MacBookPro String % ./test
Please enter a string: 
aab
["a", "a", "b"], ["aa", "b"]
min cut: 1

(base) soungmunkim@gimseongmun-ui-MacBookPro String % ./test
Please enter a string: 
"a"
["a"]
min cut: 0
 
(base) soungmunkim@gimseongmun-ui-MacBookPro String % ./test
Please enter a string: 
"ab"
["a", "b"]
min cut: 1
*/

#include <iostream>
#include <vector>
#include <string>
#include <stdbool.h>
#include <limits.h>

using namespace std;

// 복잡한 문자열도 palindrome인지 확인하는 방법 
bool palindrome(string s) {
    string cleanedString = "";
    for(char c : s) {
        if(isalnum(c)) {  // c가 알파벳 또는 숫자인지 확인
            cleanedString += tolower(c);  // 모든 문자를 소문자로 변환하여 cleanedString에 추가
        }
    }

    string reversedString = cleanedString;
    reverse(reversedString.begin(), reversedString.end());  // cleanedString을 뒤집어 reversedString에 저장

    return cleanedString == reversedString;
}

void backtrack(int start, string& s, vector<string>& path, vector<vector<string>>& result){
    if (start == s.size()){
        result.push_back(path);
    }
    // 현재 위치에서 시작하는 모든 부분 문자열에 대해
    for(int end = start; end < s.size(); end++){
        // 현재 부분 문자열이 회문이라면
        if(palindrome(s.substr(start, end-start+1))){

            // 이 부분 문자열을 path에 추가하고
            path.push_back(s.substr(start, end-start+1));
            
            // 다음 부분 문자열을 위해 재귀 호출
            backtrack(end+1, s, path, result);
            
            // 백트래킹
            path.pop_back();
        }
    }
}

void palindrome_partition(string &s){
    vector<string> path;
    vector<vector<string>> result;
    backtrack(0, s, path, result);

    // 2D array인 result를 [], [] 형태로 프린트하기
        for (size_t j = 0; j < result.size(); ++j) {
        cout << "[";
        for (size_t i = 0; i < result[j].size(); ++i) {
            cout << "\"" << result[j][i] << "\"";
            if (i != result[j].size() - 1) {  // 마지막 원소가 아닌 경우에만 쉼표를 추가
                cout << ", ";
            }
        }
        cout << "]";
        if (j != result.size() - 1) { // 마지막 행이 아닌 경우에만 쉼표와 공백을 추가
            cout << ", ";
        }
    }
    cout << endl;

     /*
     모든 partitions 돌면서 길이가 가장 작은 것 구하기
    */

    int min_len = INT_MAX;
    int partition_size;
    int min_cut;
    for (size_t i = 0; i < result.size(); i++){
        partition_size = result[i].size();
        // cout <<"partition size: " << partition_size << endl;
        if(partition_size < min_len){
            min_len = partition_size;
            // cout << "min legth: " << min_len << endl;
        }
    }
    min_cut = min_len-1;
    cout << "min cut: " << min_cut << endl;
}

// 쌍따옴표를 제거하는 함수 (문자열 사이에 있을 경우)
string removeQuotes(const string& input) {
    string result = "";
    for(char c : input) {
        if (c != '"') {  // 쌍따옴표가 아니면 결과 문자열에 추가
            result += c;
        }
    }
    return result;
}


int main(){
    string s;
    int result; 
    int min_cut;

    cout << "Please enter a string: " << endl;
    getline(cin, s);

    s = removeQuotes(s); 

    palindrome_partition(s);

    return 0;
}

File: soungmunkim/Python/String/132_PalindromePartitioning2.py
"""
[palindrome partitions 중에 가장 적은 cut 수 구하기]

Given a string s, 
partition s such that every substring of the partition is a palindrome.

Return the minimum cuts needed for a palindrome partitioning of s.

Example 1:
Input: s = "aab"
Output: 1
Explanation: The palindrome partitioning ["aa","b"] could be produced using 1 cut.

Example 2:
Input: s = "a"
Output: 0

Example 3:
Input: s = "ab"
Output: 1
"""

# 복잡한 문자열도 palindrome인지 체크함
def is_palindrome(s: str) -> bool:
    # 알파벳과 숫자만을 포함하는 새로운 문자열 생성
    cleaned_string = ''.join(char.lower() for char in s if char.isalnum())
    
    # cleaned_string을 뒤집은 문자열
    reversed_string = cleaned_string[::-1]
    
    return cleaned_string == reversed_string

# def palindrome(s: str):
#     # 문자열이 회문인지 확인하는 함수
#     return s == s[::-1]

# 모든 substring palindrome partition 찾기
def palindrome_partition(s: str):
    # 문자열의 모든 회문 파티션을 찾는 함수
    result = []
    path = []

    def backtrack(start):
        # 현재 위치가 문자열의 끝이라면, path를 결과에 추가
        if start == len(s):
            # print("path: ", path)
            result.append(list(path))
            return

        # 현재 위치에서 시작하는 모든 부분 문자열에 대해
        for end in range(start, len(s)):
            
            # 현재 부분 문자열이 회문이라면
            if is_palindrome(s[start:end+1]):
                
                # 이 부분 문자열을 path에 추가하고
                path.append(s[start:end+1])

                # 다음 부분 문자열을 위해 재귀 호출
                backtrack(end+1)
                
                # 백트래킹
                path.pop()

    backtrack(0)

    return result

def mincut(s:str):
    partitions = palindrome_partition(s)
    # 가장 길이가 작은 partitions 찾기
    min_len = float('inf')
    for i in range(len(partitions)):
        if len(partitions[i]) < min_len:
            min_len = len(partitions[i])
            
    #cut는 그 길이 -1 
    min_cut = min_len -1
    
    return min_cut

